### ISSUE TITLE
ArrayIndexOutOfBoundsException: java.lang.System in arraycopy

### ISSUE THREAD

**Issue by mhernand40** [WalletConnect/WalletConnectKotlinV2#1138]
**Describe the bug**
Seeing the following crash in production
```
java.lang.ArrayIndexOutOfBoundsException: src.length=802 srcPos=784 dst.length=16 dstPos=32 length=-16
    at java.lang.System.arraycopy(System.java)
    at org.bouncycastle.crypto.macs.Poly1305.update
    at org.bouncycastle.crypto.modes.ChaCha20Poly1305.doFinal
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptPayload-THFQ7Pg
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptPayload-THFQ7Pg
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptEnvelopeType0
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptEnvelopeType0
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encrypt
    at com.walletconnect.android.internal.common.json_rpc.domain.JsonRpcInteractor.publishJsonRpcRequest
    at com.walletconnect.android.internal.common.model.type.JsonRpcInteractorInterface$DefaultImpls.publishJsonRpcRequest$default
    at com.walletconnect.sign.engine.use_case.calls.ProposeSessionUseCase$proposeSession$2.invokeSuspend
    at com.walletconnect.sign.engine.use_case.calls.ProposeSessionUseCase$proposeSession$2.invoke
    at com.walletconnect.sign.engine.use_case.calls.ProposeSessionUseCase$proposeSession$2.invoke
    at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn
    at kotlinx.coroutines.SupervisorKt.supervisorScope
    at kotlinx.coroutines.SupervisorKt.supervisorScope
    at com.walletconnect.sign.engine.use_case.calls.ProposeSessionUseCase.proposeSession
    at com.walletconnect.sign.engine.domain.SignEngine.proposeSession
    at com.walletconnect.sign.client.SignProtocol$connect$1.invokeSuspend
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith
    at kotlinx.coroutines.DispatchedTask.run
    at kotlinx.coroutines.DispatchedTask.run
    at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run
    at kotlinx.coroutines.scheduling.TaskImpl.run
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely
    at kotlinx.coroutines.scheduling.Task.getMode$kotlinx_coroutines_core
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run
```

**SDK Version**
 - Client: Kotlin
 - Version 1.16.0

**To Reproduce**
N/A

**Device (please complete the following information):**
 - Device: Xiaomi Mi 8 Lite
 - OS: Android 10
 - Browser [OpenSea](https://play.google.com/store/apps/details?id=io.opensea&hl=en_US&gl=US) App
 - Version 2.26.2

**Comment by mhernand40** [comment]
Here is another similar stack trace:
```
java.lang.ArrayIndexOutOfBoundsException: src.length=136 srcPos=16 dst.length=16 dstPos=32 length=-16
    at java.lang.System.arraycopy(System.java)
    at org.bouncycastle.crypto.macs.Poly1305.update
    at org.bouncycastle.crypto.modes.ChaCha20Poly1305.processBytes
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptPayload-THFQ7Pg
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptPayload-THFQ7Pg
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptEnvelopeType0
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptEnvelopeType0
    at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encrypt
    at com.walletconnect.android.internal.common.json_rpc.domain.JsonRpcInteractor.publishJsonRpcRequest
    at com.walletconnect.android.internal.common.model.type.JsonRpcInteractorInterface$DefaultImpls.publishJsonRpcRequest$default
    at com.walletconnect.sign.engine.use_case.calls.DisconnectSessionUseCase$disconnect$2.invokeSuspend
    at com.walletconnect.sign.engine.use_case.calls.DisconnectSessionUseCase$disconnect$2.invoke
    at com.walletconnect.sign.engine.use_case.calls.DisconnectSessionUseCase$disconnect$2.invoke
    at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn
    at kotlinx.coroutines.SupervisorKt.supervisorScope
    at kotlinx.coroutines.SupervisorKt.supervisorScope
    at com.walletconnect.sign.engine.use_case.calls.DisconnectSessionUseCase.disconnect
    at com.walletconnect.sign.engine.domain.SignEngine.disconnect
    at com.walletconnect.sign.client.SignProtocol$disconnect$1.invokeSuspend
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith
    at kotlinx.coroutines.DispatchedTask.run
    at kotlinx.coroutines.DispatchedTask.run
    at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run
    at kotlinx.coroutines.scheduling.TaskImpl.run
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely
    at kotlinx.coroutines.scheduling.Task.getMode$kotlinx_coroutines_core
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run
```

 - Device: Samsung Galaxy A20
 - OS: Android 11
 - Browser [OpenSea](https://play.google.com/store/apps/details?id=io.opensea&hl=en_US&gl=US) App
 - Version 2.26.2

**Comment by Elyniss** [comment]
This should be fixed with version 1.17.0.

I'm closing the issue for now. Let me know if the issue prevails!

**Comment by spartan-khanhtruong** [comment]
` java.lang.ArrayIndexOutOfBoundsException: src.length=257 srcPos=32 dst.length=16 dstPos=32 length=-16
                                                                                                    	at java.lang.System.arraycopy(Native Method)
                                                                                                    	at org.bouncycastle.crypto.macs.Poly1305.update(Unknown Source:30)
                                                                                                    	at org.bouncycastle.crypto.modes.ChaCha20Poly1305.processBytes(Unknown Source:175)
                                                                                                    	at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptPayload-THFQ7Pg(ChaChaPolyCodec.kt:143)
                                                                                                    	at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encryptEnvelopeType0(ChaChaPolyCodec.kt:103)
                                                                                                    	at com.walletconnect.android.internal.common.crypto.codec.ChaChaPolyCodec.encrypt(ChaChaPolyCodec.kt:42)
                                                                                                    	at com.walletconnect.android.internal.common.json_rpc.domain.JsonRpcInteractor.publishJsonRpcRequest(JsonRpcInteractor.kt:98)
                                                                                                    	at com.walletconnect.android.internal.common.model.type.JsonRpcInteractorInterface$DefaultImpls.publishJsonRpcRequest$default(JsonRpcInteractorInterface.kt:28)
                                                                                                    	at com.walletconnect.sign.engine.use_case.calls.EmitEventUseCase$emit$2.invokeSuspend(EmitEventUseCase.kt:38)
                                                                                                    	at com.walletconnect.sign.engine.use_case.calls.EmitEventUseCase$emit$2.invoke(Unknown Source:8)
                                                                                                    	at com.walletconnect.sign.engine.use_case.calls.EmitEventUseCase$emit$2.invoke(Unknown Source:4)
                                                                                                    	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:78)
                                                                                                    	at kotlinx.coroutines.SupervisorKt.supervisorScope(Supervisor.kt:61)
                                                                                                    	at com.walletconnect.sign.engine.use_case.calls.EmitEventUseCase.emit(EmitEventUseCase.kt:30)
                                                                                                    	at com.walletconnect.sign.engine.domain.SignEngine.emit(Unknown Source:7)
                                                                                                    	at com.walletconnect.sign.client.SignProtocol$emit$1.invokeSuspend(SignProtocol.kt:304)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)`

still happens in 1.22.2
